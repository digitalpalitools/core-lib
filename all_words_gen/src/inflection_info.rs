use crate::stem_info::StemInfo;
use pls_core_extras::inflection_generator::InflectionGenerator;
use serde::Serialize;

#[derive(Serialize, Debug)]
pub struct InflectionInfo {
    pub pali1_id: i64,
    pub inflection: String,
}

#[derive(Serialize, Debug)]
pub struct InflectionsBatchInfo {
    pub inflection_infos: Vec<InflectionInfo>,
    pub inflected_forms_fetched: i64,
}

pub fn create_inflection_infos(
    stem_infos: Vec<StemInfo>,
    igen: &dyn InflectionGenerator,
) -> InflectionsBatchInfo {
    let mut ibi = InflectionsBatchInfo {
        inflection_infos: vec![],
        inflected_forms_fetched: 0,
    };

    for stem_info in stem_infos {
        let mut infs = igen.generate_all_inflections(&stem_info.pali1);
        infs.sort();
        infs.dedup();
        if infs.is_empty() {
            ibi.inflected_forms_fetched += 1;
        }

        let mut inf_infos: Vec<InflectionInfo> = infs
            .into_iter()
            .map(|inf| InflectionInfo {
                pali1_id: stem_info.row_id,
                inflection: inf,
            })
            .collect();
        ibi.inflection_infos.append(&mut inf_infos);
    }

    ibi
}

#[cfg(test)]
mod tests {
    use super::*;
    use pls_core_extras::inflection_generator::PlsInflectionGenerator;
    use pls_core_extras::logger::{NullLogger, PlsLogger};
    use pls_core_extras::sql_access::resolve_file_in_manifest_dir;

    #[test]
    fn test_create_inflection_infos() {
        let db_path = &get_db_path();
        let igen = create_inflection_generator(db_path, &NullLogger {});
        let stems = vec![
            StemInfo {
                row_id: 7,
                pali1: "ababa 1".to_string(),
            },
            StemInfo {
                row_id: 8,
                pali1: "ababa 2".to_string(),
            },
            StemInfo {
                row_id: 32,
                pali1: "abbahe".to_string(),
            },
        ];
        let infl_infos = create_inflection_infos(stems, &igen);

        insta::assert_yaml_snapshot!(infl_infos);
    }

    fn create_inflection_generator<'a>(
        db_path: &'a str,
        logger: &'a dyn PlsLogger,
    ) -> PlsInflectionGenerator<'a> {
        let igen = PlsInflectionGenerator::new(
            "en",
            env!("CARGO_PKG_VERSION"),
            env!("CARGO_PKG_NAME"),
            db_path,
            logger,
        )
        .expect("unexpected test setup failure");

        igen
    }

    fn get_db_path() -> String {
        resolve_file_in_manifest_dir("inflections.db")
            .expect("must exist")
            .as_path()
            .to_str()
            .expect("must exist")
            .to_string()
    }
}
