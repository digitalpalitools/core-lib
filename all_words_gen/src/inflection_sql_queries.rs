use crate::inflection_info::InflectionsBatchInfo;
use serde::Serialize;

#[derive(Serialize, Debug)]
pub struct InflectionSqlQueryBatchInfo {
    pub inflection_sql_queries: Vec<String>,
    pub inflected_forms_fetched: i64,
}

pub fn create_inflection_sql_queries(
    inflection_infos: InflectionsBatchInfo,
) -> InflectionSqlQueryBatchInfo {
    let mut sql_queries: Vec<String> = inflection_infos
        .inflection_infos
        .into_iter()
        .map(|inf| {
            format!(
                "INSERT INTO _all_words (inflection, stem_id) VALUES ('{}', {})",
                inf.inflection, inf.pali1_id
            )
        })
        .collect();

    // NOTE: Add transactions to make insertion faster https://stackoverflow.com/a/3852082/6196679
    let mut inflection_sql_queries = vec!["BEGIN".to_string()];
    inflection_sql_queries.append(&mut sql_queries);
    inflection_sql_queries.push("END".to_string());

    InflectionSqlQueryBatchInfo {
        inflection_sql_queries,
        inflected_forms_fetched: inflection_infos.inflected_forms_fetched,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::inflection_info::InflectionInfo;

    #[test]
    fn test_create_inflection_infos() {
        let inflection_infos = InflectionsBatchInfo {
            inflected_forms_fetched: 10005,
            inflection_infos: vec![
                InflectionInfo {
                    pali1_id: 102,
                    inflection: "102-1".to_string(),
                },
                InflectionInfo {
                    pali1_id: 102,
                    inflection: "102-2".to_string(),
                },
                InflectionInfo {
                    pali1_id: 103,
                    inflection: "103-1".to_string(),
                },
            ],
        };

        let inflected_forms_fetched = inflection_infos.inflected_forms_fetched;
        let isqbi = create_inflection_sql_queries(inflection_infos);

        assert_eq!(inflected_forms_fetched, isqbi.inflected_forms_fetched);
        insta::assert_yaml_snapshot!(isqbi);
    }
}
