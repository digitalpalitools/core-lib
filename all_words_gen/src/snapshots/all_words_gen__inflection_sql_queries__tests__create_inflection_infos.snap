---
source: all_words_gen/src/inflection_sql_queries.rs
expression: isqbi

---
inflection_sql_queries:
  - BEGIN
  - "INSERT INTO _all_words (inflection, stem_id) VALUES ('102-1', 102)"
  - "INSERT INTO _all_words (inflection, stem_id) VALUES ('102-2', 102)"
  - "INSERT INTO _all_words (inflection, stem_id) VALUES ('103-1', 103)"
  - END
inflected_forms_fetched: 10005

